File: src/rails/game/state/IntegerChange.java
package rails.game.state;

/**
 * Change associated with IntegerState
 */
final class IntegerChange extends Change {

    private final IntegerState state;
    private final int newValue;
    private final int oldValue;
    
    IntegerChange(IntegerState state, int newValue) {
        this.state = state;
        this.newValue = newValue;
        this.oldValue = state.value();
        super.init(state);
    }

    @Override
    public void execute() {
        state.change(newValue);
    }

    @Override
    public void undo() {
        state.change(oldValue);
    }

    @Override
    public IntegerState getState() {
        return state;
    }

    @Override
    public String toString() {
        return "Change for " + state + ": From " + oldValue + " to " + newValue; 
    }

}

File: dotnet/RailsLib.Net/Net/Game/State/IntegerChange.cs
using System;


/**
 * Change associated with IntegerState
 */

namespace GameLib.Net.Game.State
{
    public sealed class IntegerChange : Change
    {
        private readonly IntegerState state;
        private readonly int newValue;
        private readonly int oldValue;

        public IntegerChange(IntegerState state, int newValue)
        {
            this.state = state;
            this.newValue = newValue;
            this.oldValue = state.Value;
            base.Init(state);
        }

        override public void Execute()
        {
            state.Change(newValue);
        }

        override public void Undo()
        {
            state.Change(oldValue);
        }

        override public GameState GameState
        {
            get
            {
                return state;
            }
        }

        override public string ToString()
        {
            return "Change for " + state + ": From " + oldValue + " to " + newValue;
        }
    }
}
